torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -344,8 +357,12 @@ class ChunkDataset final
 
     chunk_reader_.reset();
 
-    size_t chunks_to_load = chunk_reader_.chunk_count();
-    chunk_sampler_.reset(chunks_to_load);
+    // reset the chunk sampler.
+    chunk_sampler_->reset(nullopt);
+
+    // In distributed training, local chunk count could be different to total
+    // chunks availble. Chunk sampler holds the truth.
+    size_t chunks_to_load = chunk_sampler_->local_sample_count();
```

Comment:
No, no, no, this is a clear leakage of abstraction. We should make it such that the dataset stops loading as soon as the sampler runs out of indices. The size is completely unnecessary here.